capsules/src/proximity.rs

```diff
@@ -0,0 +1,310 @@
+        let mut num_commands: u8 = 0;
+
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                if app.subscribed {
+                    num_commands += 1;
+                }
+            });
+        }
+        if num_commands == 1 {
+            self.run_next_command();
+        }
+
+        ReturnCode::SUCCESS
+    }
+
+    fn run_next_command(&self) -> ReturnCode {
+        let mut break_flag: bool = false;
+
+        // Find and run another command
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                if app.subscribed {
+                    // run it
+                    match app.enqueued_command_type {
+                        ProximityCommand::ReadProximity => {
+                            self.call_driver(app.enqueued_command_type, 0, 0);
+                        }
+                        ProximityCommand::ReadProximityOnInterrupt => {
+                            let t: Thresholds = self.find_thresholds();
+                            self.call_driver(
+                                app.enqueued_command_type,
+                                t.lower as usize,
+                                t.upper as usize,
+                            );
+                        }
+                        _ => {}
+                    }
+
+                    break_flag = true;
+                }
+            });
+
+            if break_flag {
+                break;
+            }
+        }
+
+        ReturnCode::SUCCESS
+    }
+
+    fn find_thresholds(&self) -> Thresholds {
+        // Get the lowest upper prox and highest lower prox of all subscribed apps
+        // With the IC thresholds set to these two values, we ensure to never miss an interrupt-causing proximity value for any of the
+        // apps
+        let mut highest_lower_proximity: u8 = 0;
+        let mut lowest_upper_proximity: u8 = 255;
+
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                if (app.lower_proximity > highest_lower_proximity) && app.subscribed {
+                    highest_lower_proximity = app.lower_proximity;
+                }
+                if (app.upper_proximity < lowest_upper_proximity) && app.subscribed {
+                    lowest_upper_proximity = app.upper_proximity;
+                }
+            });
+        }
+
+        // return values
+        Thresholds {
+            lower: highest_lower_proximity,
+            upper: lowest_upper_proximity,
+        }
+    }
+
+    fn call_driver(&self, command: ProximityCommand, arg1: usize, arg2: usize) -> ReturnCode {
+        match command {
+            ProximityCommand::ReadProximity => self.driver.read_proximity(),
+            ProximityCommand::ReadProximityOnInterrupt => self
+                .driver
+                .read_proximity_on_interrupt(arg1 as u8, arg2 as u8),
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
```

Comment:
Having this as its own function seems unnecessary, and passing arg1/arg2 as usizes even though they are really optional arguments is not particularly "rusty". It probably makes more sense to replace calls to this function with the appropriate calls to `self.driver.read_proximity()` or `self.driver.read_proximity_on_interrupt()`

Reply:
A good suggestion.  I have addressed this.